### PR TITLE
docker: add Dockerfile

### PR DESCRIPTION
**Pull request by Winterflower**
This is a draft of the Dockerfile I am working on for https://github.com/tmc/pqstream/issues/12
Currently the image build if failing, because of an issue with go get. The output from docker build is below
```
# cd .; git clone https://github.com/tmq/pqstream /go/src/github.com/tmq/pqstream
Cloning into '/go/src/github.com/tmq/pqstream'...
fatal: could not read Username for 'https://github.com': terminal prompts disabled
package github.com/tmq/pqstream/cmd/pqsd: exit status 128
```
I will debug! :)

### CODE REVIEW

**Comment by tmc** [comment]
@Winterflower this is awesome! I just rebased/merged master which adds image building to the circle job (and set circle to build forks).

One thing to consider is that the ./check_generated_code script is part of the normal ci job which exits non-zero if the generated code appears out of date -- you may be able to re-use it in image building. Another option is to not publish images unless both parts of the circle job complete (build+test, docker image build). Feel free to hop in #pqstream on gophers slack if you want to discuss more.

**Comment by Winterflower** [comment]
hi @tmc ! 
unfortunately the image build is failing (same issue as I have locally - the go get command fails during the build). The solution is to add a few lines to .gitconfig and some ssh keys config. I'll add it to the Dockerfile.
:) I think I need an invite to join the gopher slack (per the message at the bottom of this https://gophers.slack.com/)

Edit: will take a look at check_generated_code

**Comment by tmc** [comment]
I haven't looked closely, maybe drop the -u?

Invite at https://invite.slack.golangbridge.org/

**Comment by Winterflower** [comment]
``docker build`` now passes for the Dockerfile in this branch :) 
I'll take a look at running check_generated_code in the docker image build phase

**Review by tmc — COMMENTED** [review]
This is awesome!

As is I think this will build whatever is current which is fine but I think it would be better if we used COPY to copy code to /go/src/github.com/tmc/pqstream and then built the binaries via a 'go install' run. Thoughts?

**Comment by tmc** [comment]
@Winterflower can we move this to use COPY so branches don't build master?

**Comment by Winterflower** [comment]
Yup, got you, will make change

**Comment by Winterflower** [comment]
@tmc (just to clarify, to build a docker image for the build's branch, we want the following)
1. git clone $REPO && cd <repo>
2. git checkout $BRANCH (which is passed from CIRCLE CI env vars to docker build --build-args)
4. COPY <repo> $GOPATH/src/github.com/tmc/
5. go install 

how will we manage the deps?

for the record, (if I understood you correctly), I tried to build an image from sth like this and am hitting the dependency issue. 
As far as I can tell there is no way to pass a flag to go get to make it checkout a specific branch, not master.

```
FROM golang:alpine

RUN apk add --update git 

RUN git clone https://github.com/tmc/pqstream \
    && cd pqstream \
    && git checkout ${BRANCH}

ENV PKG_PATH $GOPATH/src/github.com/tmc/pqstream 

COPY . $PKG_PATH

RUN cd $PKG_PATH \
     && go install .              
```

File: Dockerfile [hunk 1]
```diff
@@ -1,3 +1,17 @@
...
+    && go get -v github.com/tmc/pqstream/cmd/pqsd
+
+RUN go install github.com/tmc/pqstream/cmd/pqs \
+    && go install github.com/tmc/pqstream/cmd/pqsd
+
+ENV PATH $PATH:$GOPATH/bin
+
+EXPOSE 7000
+
+ENTRYPOINT ["pqsd"]
```
Comment (tmc) [inline Dockerfile:17]: entrypoint vs CMD implies we wouldn't be able to run the client (pqs) via the same image

**Review by tmc — APPROVED** [review]
We can fix that up later. LGTM!

File: Dockerfile (hunk 1 above)
Comment (Winterflower) [reply Dockerfile:17]: ah sorry about this - I did remember you mentioned CMD on Slack, but after reading some entrypoint vs cmd docs, I decided to go with entrypoint, b/c it seemed that ti would be possible for users to override the command in CMD - I was prob. confused!

**Pull request merged**